zebra-chain/src/primitives/redpallas/signing_key.rs

```diff
@@ -30,10 +38,88 @@ impl<T: SigType> TryFrom<[u8; 32]> for SigningKey<T> {
 
         if maybe_sk.is_some().into() {
             let sk = maybe_sk.unwrap();
-            let pk = VerificationKey::from(&sk);
+            let pk = VerificationKey::from_scalar(&sk);
             Ok(SigningKey { sk, pk })
         } else {
             Err(Error::MalformedSigningKey)
         }
     }
 }
+
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+struct SerdeHelper([u8; 32]);
+
+impl<T: SigType> TryFrom<SerdeHelper> for SigningKey<T> {
+    type Error = Error;
+
+    fn try_from(helper: SerdeHelper) -> Result<Self, Self::Error> {
+        helper.0.try_into()
+    }
+}
+
+impl<T: SigType> From<SigningKey<T>> for SerdeHelper {
+    fn from(sk: SigningKey<T>) -> Self {
+        Self(sk.into())
+    }
+}
+
+impl SigningKey<SpendAuth> {
+    /// Randomize this public key with the given `randomizer`.
+    pub fn randomize(&self, randomizer: &pallas::Scalar) -> SigningKey<SpendAuth> {
+        let sk = self.sk + randomizer;
+        let pk = VerificationKey::from_scalar(&sk);
+        SigningKey { sk, pk }
+    }
+}
+
+impl<T: SigType> SigningKey<T> {
+    /// Generate a new signing key.
+    pub fn new<R: RngCore + CryptoRng>(mut rng: R) -> SigningKey<T> {
+        let sk = {
+            let mut bytes = [0; 64];
+            rng.fill_bytes(&mut bytes);
+            pallas::Scalar::from_bytes_wide(&bytes)
+        };
+        let pk = VerificationKey::from_scalar(&sk);
+        SigningKey { sk, pk }
+    }
+
+    /// Create a signature of type `T` on `msg` using this `SigningKey`.
+    ///
+    /// https://zips.z.cash/protocol/nu5.pdf#concretereddsa
+    // Similar to signature::Signer but without boxed errors.
+    pub fn sign<R: RngCore + CryptoRng>(&self, mut rng: R, msg: &[u8]) -> Signature<T> {
+        use super::HStar;
+
+        // RedDSA.GenRandom:() → R RedDSA.Random
+        // Choose a byte sequence uniformly at random of length
+        // (\ell_H + 128)/8 bytes.  For RedPallas this is (512 + 128)/8 = 80.
+        let random_bytes = {
+            let mut bytes = [0; 80];
+            rng.fill_bytes(&mut bytes);
+            bytes
+        };
+
+        let nonce = HStar::default()
+            .update(&random_bytes[..])
+            .update(&self.pk.bytes.bytes[..]) // XXX ugly
+            .update(msg)
+            .finalize();
+
+        let r_bytes = pallas::Affine::from(T::basepoint() * nonce).to_bytes();
+
+        let c = HStar::default()
+            .update(&r_bytes[..])
+            .update(&self.pk.bytes.bytes[..]) // XXX ugly
```

Comment:
Try:
```suggestion
            .update(self.pk.bytes.bytes.as_slice())
            .update(msg)
            .finalize();

        let r_bytes = pallas::Affine::from(T::basepoint() * nonce).to_bytes();

        let c = HStar::default()
            .update(&r_bytes[..])
            .update(self.pk.bytes.bytes.as_slice())
```

Reply:
Ah I would but it's not available on stable, even with the feature flag

Reply:
Oh right that's the one as_slice impl that isn't stable.